gcloud-java-bigquery/src/main/java/com/google/cloud/bigquery/Job.java

```diff
@@ -144,14 +163,51 @@ public boolean isDone() {
   }
 
   /**
+   * Waits until this job completes its execution, either failing or succeeding. If the job does not
+   * exist, this method returns without executing any method of the provided callback. If the job
+   * completed successfully the {@link CompletionCallback#success(Job)} method is called. If the job
+   * completed with errors the {@link CompletionCallback#error(BigQueryError, List)} method is
+   * called.
+   * <pre> {@code
+   * job.whenDone(new CompletionCallback() {
+   *   void success(Job job) {
+   *     // completed successfully
+   *   }
+   *
+   *   void error(BigQueryError error, List<BigQueryError> executionErrors) {
+   *     // handle error
+   *   }
+   * });}</pre>
+   *
+   * @throws BigQueryException upon failure
+   * @throws InterruptedException if the current thread gets interrupted while waiting for the job
+   *     to complete
+   */
+  public void whenDone(CompletionCallback callback) throws InterruptedException {
```

Comment:
I don't like this method either and I don't like the idea of writing `Thread.sleep` somewhere in our code. But if we see this as a way to avoid the user to write boilerplate code then it might make sense.

> I think it would have been much more useful if it was not blocking and we could return a "listenable future".

Yes but as long as we stick to Java 7 this is not doable. Also the whole BigQuery/Compute API is synchronous, such a change would make sense when providing all other methods in an async flavor. 

> 1) It is not clear that this method is blocking.Though Java doc says "Waits until..." I would be more explicit saying "this method will block until..." and also change the name of the method to something like waitFor.

Ok to me

> 2) Existing the method without a callback and without indicating that a callback was not called is error prone

Non-existing operations/jobs are a very unlikely case. I believe that having a dedicated method in the callback is not worth it. Maybe the return value is the less invasive way of handling this case, but it's hardly intuitive.

> 3) There is actually another exit point in this case and that could be that the RPC for isDone or reload failed, which similar to (2) makes the callback usage more complex. A solution for that would be to never fail 
calling this method and to indicate any failures (in a distinct way) in the callback.

I don't feel too strange that the method throws isDone or reload failed. The method is blocking and the callback is related to the operation status, not the method's error condition, so exceptions make sense to me.

> 4) 500L sampling rate may be too course or to grain depending on the job and the client needs. Maybe we could provide a var-arg way to tune it.

Vararg options to tune this sound good to me even though I don't expect much users to be interested in tuning this. BTW 500 if I recall correctly is the value adopted by node guys.